File: backend/internal/engine/timemanager.go
package engine

import "caro-ai-pvp/internal/domain"

type TimeAllocation struct {
	SoftBoundMs int64
	HardBoundMs int64
	OptimalMs   int64
}

func AllocateTime(timeRemainingMs int64, incrementMs int64, moveNumber int) TimeAllocation {
	phaseDivisor := domain.TimePhaseDivisorEarly
	if moveNumber > domain.TimePhaseSwitchMove {
		phaseDivisor = domain.TimePhaseDivisorLate
	}

	baseMs := float64(timeRemainingMs) / phaseDivisor
	incContrib := float64(incrementMs) * domain.TimeIncContribFactor

	optimal := int64(baseMs + incContrib)
	if optimal < domain.TimeMinOptimalMs {
		optimal = domain.TimeMinOptimalMs
	}

	maxTime := int64(float64(timeRemainingMs) * domain.TimeMaxFraction)
	if optimal > maxTime {
		optimal = maxTime
	}

	hardBound := int64(float64(optimal) * domain.TimeHardBoundMultiplier)
	buffer := int64(float64(timeRemainingMs) * domain.TimeBufferFraction)
	if buffer < domain.TimeMinBufferMs {
		buffer = domain.TimeMinBufferMs
	}
	hardBound += buffer
	if hardBound > timeRemainingMs-domain.TimeReserveMs {
		hardBound = timeRemainingMs - domain.TimeReserveMs
	}
	// Any live clock still deserves a usable budget: a zero or negative hard
	// bound makes the search abort instantly and fall back to move ordering.
	if timeRemainingMs > 0 && hardBound < domain.TimeMinBufferMs {
		hardBound = domain.TimeMinBufferMs
	}
	if hardBound < 0 {
		hardBound = 0
	}

	softBound := int64(float64(optimal) * domain.TimeSoftBoundFraction)

	return TimeAllocation{
		SoftBoundMs: softBound,
		HardBoundMs: hardBound,
		OptimalMs:   optimal,
	}
}

File: backend/src/Caro.Engine/TimeManager.cs
using Caro.Domain;

namespace Caro.Engine;

public readonly record struct TimeAllocation(long SoftBoundMs, long HardBoundMs, long OptimalMs);

public static class TimeManager
{
    public static TimeAllocation AllocateTime(long timeRemainingMs, long incrementMs, int moveNumber)
    {
        double phaseDivisor = Constants.TimeManagement.PhaseDivisorEarly;
        if (moveNumber > Constants.TimeManagement.PhaseSwitchMove)
        {
            phaseDivisor = Constants.TimeManagement.PhaseDivisorLate;
        }

        double baseMs = timeRemainingMs / phaseDivisor;
        double incContrib = incrementMs * Constants.TimeManagement.IncContribFactor;

        long optimal = (long)(baseMs + incContrib);
        if (optimal < Constants.TimeManagement.MinOptimalMs)
        {
            optimal = Constants.TimeManagement.MinOptimalMs;
        }

        long maxTime = (long)(timeRemainingMs * Constants.TimeManagement.MaxFraction);
        if (optimal > maxTime)
        {
            optimal = maxTime;
        }

        long hardBound = (long)(optimal * Constants.TimeManagement.HardBoundMultiplier);
        long buffer = (long)(timeRemainingMs * Constants.TimeManagement.BufferFraction);
        if (buffer < Constants.TimeManagement.MinBufferMs)
        {
            buffer = Constants.TimeManagement.MinBufferMs;
        }
        hardBound += buffer;
        if (hardBound > timeRemainingMs - Constants.TimeManagement.ReserveMs)
        {
            hardBound = timeRemainingMs - Constants.TimeManagement.ReserveMs;
        }
        // Any live clock still deserves a usable budget: a zero or negative
        // hard bound makes the search abort instantly and fall back to move
        // ordering.
        if (timeRemainingMs > 0 && hardBound < Constants.TimeManagement.MinBufferMs)
        {
            hardBound = Constants.TimeManagement.MinBufferMs;
        }
        if (hardBound < 0)
        {
            hardBound = 0;
        }

        long softBound = (long)(optimal * Constants.TimeManagement.SoftBoundFraction);

        return new TimeAllocation(softBound, hardBound, optimal);
    }
}
